kprobe_queue.c

```diff
@@ -787,59 +787,128 @@ parse_data_offset(void)
 	return (data_offset);
 }
 
+#define TOKSZ	256
+
+static int
+kprobe_exp_split(const char *exp1, char left[TOKSZ], char *op, char right[TOKSZ])
+{
+	int	 paren_depth;
+	char	 exp[1024];
+	char	*p, *start_left, *start_right;
+
+	if (strlcpy(exp, exp1, sizeof(exp)) >= sizeof(exp)) {
+		qwarnx("expression too long");
+		return (-1);
+	}
+	paren_depth = 0;
+	start_left = start_right = NULL;
+	*op = 0;
+
+	for (p = exp; *p != 0; p++) {
+		switch (*p) {
+		case '(':
+			if (++paren_depth == 1)
+				start_left = p + 1;
+			break;
+		case '+':	/* FALLTHROUGH */
+		case '-':	/* FALLTHROUGH */
+		case '*':	/* FALLTHROUGH */
+			if (paren_depth > 1)
+				break;
+			if (*op != 0) {
+				qwarnx("multiple operators");
+				return (-1);
+			}
+			*op = *p;
+			*p = 0;
+			start_right = p + 1;
+			break;
+		case ')':
+			if (--paren_depth == 0)
+				*p = 0;
+			if (paren_depth < 0) {
+				qwarnx("unbalanced parenthesis");
+				return (-1);
+			}
+			break;
+		default:
+			break;
+		}
+	}
+	if (*op == 0) {
+		qwarnx("no operator");
+		return (-1);
+	}
+	if (start_left == NULL || start_right == NULL) {
+		qwarnx("syntax error");
+		return (-1);
+	}
+	if (strlcpy(left, start_left, TOKSZ) >= TOKSZ) {
+		qwarnx("left token overflow");
+		return (-1);
+	}
+	if (strlcpy(right, start_right, TOKSZ) >= TOKSZ) {
+		qwarnx("right token overflow");
+		return (-1);
+	}
+
```

Comment:
Add length checks for left and right?

I think mistakes like `(+16)` or `(16+)` will result in hard to understand error messages later.

Reply:
you mean is left or right is empty?